python-sdk/nuscenes/utils/kitti.py

```diff
@@ -0,0 +1,526 @@
+                    if set(expected_tokens) != set(tokens):
+                        warnings.warn('Split {} is missing tokens for the {} table.'.format(split, table))
+
+    @staticmethod
+    def standardize_sample_token(token: str) -> Tuple[str, str]:
+        """
+        Convert sample token into standard KITTI folder and local filename format.
+        :param token: KittiDB unique id.
+        :return: folder (ex. train, val, test), filename (ex. 000001)
+        """
+        folder, filename = token.split('_')
+        return folder, filename
+
+    @staticmethod
+    def parse_label_line(label_line) -> dict:
+        """
+        Parses single line from label file into a dict. Boxes are in camera frame. See KITTI devkit for details and
+        http://www.cvlibs.net/datasets/kitti/setup.php for visualizations of the setup.
+        :param label_line: Single line from KittiDB label file.
+        :return: Dictionary with all the line details.
+        """
+
+        parts = label_line.split(' ')
+        output = {
+            'name': parts[0].strip(),
+            'xyz_camera': (float(parts[11]), float(parts[12]), float(parts[13])),
+            'wlh': (float(parts[9]), float(parts[10]), float(parts[8])),
+            'yaw_camera': float(parts[14]),
+            'bbox_camera': (float(parts[4]), float(parts[5]), float(parts[6]), float(parts[7])),
+            'truncation': float(parts[1]),
+            'occlusion': float(parts[2]),
+            'alpha': float(parts[3])
+        }
+
+        return output
+
+    @staticmethod
+    def box_nuscenes_to_kitti(box: Box, velo_to_cam_rot: Quaternion,
+                              velo_to_cam_trans: np.ndarray,
+                              r0_rect: Quaternion,
+                              kitti_to_nu_lidar_inv: Quaternion = Quaternion(axis=(0, 0, 1), angle=np.pi / 2).inverse) \
+            -> Box:
+        """
+        Transform from nupylib (nuScenes lidar frame) to KITTI reference frame.
+        :param box: Instance in nuScenes lidar frame.
+        :param velo_to_cam_rot: Quaternion to rotate from lidar to camera frame.
+        :param velo_to_cam_trans: <np.float: 3>. Translate from lidar to camera frame.
+        :param r0_rect: Quaternion to rectify camera frame.
+        :param kitti_to_nu_lidar_inv: Quaternion to rotate nuScenes to KITTI LIDAR.
+        :return: Box instance in KITTI reference frame.
+        """
+        # Copy box to avoid side-effects.
+        box = box.copy()
+
+        # Rotate to KITTI lidar.
+        box.rotate(kitti_to_nu_lidar_inv)
+
+        # Transform to KITTI camera.
+        box.rotate(velo_to_cam_rot)
+        box.translate(velo_to_cam_trans)
+
+        # Rotate to KITTI rectified camera.
+        box.rotate(r0_rect)
+
+        # KITTI defines the box center as the bottom center of the object.
+        # We use the true center, so we need to adjust half height in y direction.
+        box.translate(np.array([0, box.wlh[2] / 2, 0]))
```

Comment:
This reads a bit strange / hacky to me. The Box class is defined to have the center is the actual center. So by moving the center, the box instance is no longer valid according to the Box class. I think the right way would be to create a KITTIBox class where the center is defined as the bottom-center per KITTI convention. But on the other hand, it may be OK in these scripts.

Reply:
Yes, it is not ideal, but for this not widely used script I think it is okay.